### PR TITLE
Reworked server synchronization

### PR DESCRIPTION
**Pull request by tolusha**
Signed-off-by: Anatolii Bazko <abazko@redhat.com>

### What does this PR do?
* Reworked server synchronization

### Screenshot/screencast of this PR

N/A

### What issues does this PR fix or reference?

N/A

### How to test this PR?

N/A

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [x] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [Custom resource definition file is up to date](https://github.com/eclipse-che/che-operator/blob/main/README.md#updating-custom-resource-definition-file)
- [ ] [Nightly OLM bundle is up to date](https://github.com/eclipse-che/che-operator/blob/main/README.md#update-nightly-olm-bundle)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: pkg/deploy/checluster.go [hunk 1]
```diff
@@ -13,28 +13,74 @@ package deploy
...
+		deployContext.CheCluster.Status.HelpLink = helpLink
+		if err := UpdateCheCRStatus(deployContext, "status: HelpLink", message); err != nil {
+			return err
+		}
 	}
-	logrus.Infof("Custom resource %s updated", deployContext.CheCluster.Name)
 	return nil
 }
+
+func ReloadCheCluster(deployContext *DeployContext) error {
```
Comment (mmorhun) [inline pkg/deploy/checluster.go:81]: Reload and just Che cluster sounds ambiguous. Maybe `ReloadCheClusterCR` ?

File: pkg/deploy/server/server.go [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+
+	done, err = s.UpdateCheVersion()
+	if !done {
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (s *Server) SyncService() (bool, error) {
```
Comment (mmorhun) [inline pkg/deploy/server/server.go:111]: which service? Maybe we should reflect it in the name?

File: pkg/deploy/server/server.go [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+	if s.deployContext.CheCluster.Spec.Server.CheDebug == "true" {
+		portName = append(portName, "debug")
+		portNumber = append(portNumber, deploy.DefaultCheDebugPort)
+	}
+
+	spec := deploy.GetServiceSpec(s.deployContext, deploy.CheServiceName, portName, portNumber, s.component)
+	return deploy.Sync(s.deployContext, spec, deploy.ServiceDefaultDiffOpts)
+}
+
+func (s Server) ExposeEndpoint() (bool, error) {
```
Comment (mmorhun) [inline pkg/deploy/server/server.go:129]: The same as above (and for methods below), which endpoint? Please think of better name.

File: pkg/deploy/server/server_configmap.go [hunk 4]
```diff
@@ -87,23 +87,15 @@ type CheConfigMap struct {
...
-	}
-	return deploy.SyncConfigMapDataToCluster(deployContext, CheConfigMapName, data, deploy.DefaultCheFlavor(deployContext.CheCluster))
-}
-
 // GetCheConfigMapData gets env values from CR spec and returns a map with key:value
 // which is used in CheCluster ConfigMap to configure CheCluster master behavior
-func GetCheConfigMapData(deployContext *deploy.DeployContext) (cheEnv map[string]string, err error) {
-	cheHost := deployContext.CheCluster.Spec.Server.CheHost
-	keycloakURL := deployContext.CheCluster.Spec.Auth.IdentityProviderURL
+func (s *Server) getConfigMapData() (cheEnv map[string]string, err error) {
```
Comment (mmorhun) [inline pkg/deploy/server/server_configmap.go:92]: Comment above says it should be `GetCheConfigMapData`

File: pkg/controller/che/che_controller.go [hunk 5]
```diff
@@ -236,7 +231,7 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
 		return err
 	}
 
-	if isOpenShift {
+	if util.IsOpenShift {
```
Comment (mmorhun) [inline pkg/controller/che/che_controller.go:234]: I missed this previous time and maybe that is not for current PR, but getting a context value from `util` look pretty weird. To me, it should be in some sort of context.

**Review by mmorhun — APPROVED** [review]

File: pkg/controller/che/che_controller.go [hunk 6]
```diff
@@ -657,104 +575,35 @@ func (r *ReconcileChe) Reconcile(request reconcile.Request) (reconcile.Result, e
...
-			deployContext.CheCluster.Spec.Server.CheServerIngress,
-			cheFlavor)
-		if !done {
-			logrus.Infof("Waiting on ingress '%s' to be ready", cheFlavor)
-			if err != nil {
-				logrus.Error(err)
+	// create and provision Keycloak related objects
+	if !instance.Spec.Auth.ExternalIdentityProvider {
+		provisioned, err := identity_provider.SyncIdentityProviderToCluster(deployContext)
+		if !tests {
```
[the remark this answers is not in the document]
Comment (tolusha) [reply pkg/controller/che/che_controller.go:592]: ok

File: pkg/controller/che/che_controller.go (hunk 5 above)
Comment (tolusha) [reply pkg/controller/che/che_controller.go:234]: Sure, It is in my todo list.

**Comment by tolusha** [comment]
@mmorhun 
I've addressed all your remarks.

**Pull request merged**